### ISSUE TITLE
How to place a postOnly order with Deribit?

### ISSUE THREAD

**Issue by Tezeug** [ccxt/ccxt#5159]
- OS: Kubuntu 19.04 64-bit
- Programming Language version: Python 3.7.3
- CCXT version: 1.18.557
- Exchange: DeriBit
- Method: params={"postOnly": True}

Hello. I am trying to post a limit order on the DeriBit exchange with the postOnly parameter but I must be doing something wrong because it does not work. I have tried to add this into the createOrder command but when I fetch the order and let it print, it shows 'postOnly': False, ...

params={"postOnly": True}

Thanks for help.
I have also tried the 'execInst': 'ParticipateDoNotInitiate' but still has not worked.

**Comment by kroitor** [comment]
> Hello. I am trying to post a limit order on the DeriBit exchange with the postOnly parameter but I must be doing something wrong because it does not work. 

First, you need to answer yourself: does Deribit support the `postOnly` exchange-specific parameter at all? In order to answer that question, you need to take a brief look into their docs (the links to their docs are all here: https://github.com/ccxt/ccxt/wiki/Exchange-Markets). 

- https://docs.deribit.com/rpc-endpoints.html#buy

↑ From that doc it follows that the actual parameter name is `post_only`. So, you params have to be:

```Python
params={"post_only": True}
```

Does that resolve the issue for you?

**Comment by kroitor** [comment]
The `execInst` won't help in this case, according to their docs:

> Defines trigger type, required for "stop_limit" and "stop_narket" order type, possible values "index_price", "mark_price", "last_price" (Only valid for stop orders).

"Stop_narket", lol.

**Comment by Tezeug** [comment]
Forgot to mention. I have tried this too and does not work. The type on the website also keeps showing just Limit. Btw. yes. it supports it.

**Comment by kroitor** [comment]
> Forgot to mention. I have tried this too and does not work.

Yep, however, "does not work" does not really work for a proper description of the problem, unfortunately )

So, in this case we need you to follow very precisely these guides here (extremely precisely, literally to a word):

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-submit-an-issue

And provide the full code snippet to reproduce it, including the instantiation and the call in question, and also provide the `verbose` request/response pair for the call in question... Something like:

```Python
import ccxt
exchange = ccxt.deribit ({
    'apiKey': 'YOUR_API_KEY',
    'secret': 'YOUR_SECRET',
    'enableRateLimit': True,
})
exchange.load_markets()

amount = YOUR_AMOUNT
price = YOUR_PRICE
symbol = YOUR_SYMBOL
type = YOUR_TYPE
side = YOUR_SIDE

params={
    'post_only': True,
}

exchange.verbose = True  # ←------ !!!
exchange.create_order(symbol, type, side, amount, price, params)
```

↑ ... and paste the output here using formatting and markup rules **precisely as described in the guidelines posted above**.

**Comment by Tezeug** [comment]
Just a note. I have tried this and it has showed this error.

Exchange.privatePostBuy({"filter":{'instrument':S1, 'type':T1, 'amount':AA1, 'price':7500, 'post_only':'True'}})

  raise exceptions[error](feedback)
ccxt.base.errors.AuthenticationError: deribit {"usOut":........,"usIn":...........,"usDiff":0,"testnet":true,"success":false,"message":"authorization_required","error":10000}

**Comment by Tezeug** [comment]
Look at it, still postOnly: false.

Response: POST https://test.deribit.com/api/v1/private/buy 200 {'Server': 'nginx/1.15.12', 'Date': 'TueContent-Type': 'application/json', 'Content-Length': '316', 'Connection': 'keep-alive', 'content-encodi-Security': 'max-age=15768000', 'X-Frame-Options': 'SAMEORIGIN'} {"testnet":true,"usOut":15584410942462Diff":1451,"success":true,"result":{"order":{"orderId":"2380538411","type":"limit","instrument":"BTC-PEice":7500.0,"label":"","amount":500.0,"quantity":50,"filledQuantity":0,"filledAmount":0.0,"avgPrice":0.558441094241,"lastUpdate":1558441094241,"state":"open","postOnly":false,"reduceOnly":false,"api":true,"ShowAmount":500,"adv":false},"trades":[]},"message":""}

**Comment by Tezeug** [comment]
import ccxt
exchange = ccxt.deribit ({
    'apiKey': '',
    'secret': '',
    'enableRateLimit': True,
    "urls": {"api": "https://test.deribit.com"}
})
exchange.load_markets()

amount = 50
price = 7500
symbol = 'BTC-PERPETUAL'
type = 'limit'
side = 'buy'

params={
    'post_only': True
}

exchange.verbose = True  # ←------ !!!
exchange.create_order(symbol, type, side, amount, price, params)

**Comment by Tezeug** [comment]
When I tick the Post-Only option on the website and place an order it shows Type Post Only|Limit.

**Comment by kroitor** [comment]
It says, that is it the AuthenticationError, means something is wrong with the keys or with the signing procedure. Have you looked through other issue related to Deribit? Looks like it has been broken for a while: 

- https://github.com/ccxt/ccxt/issues/4833
- https://github.com/ccxt/ccxt/issues/3780#issuecomment-454090561

Let us know if that doesn't answer the question. I suggest to move this discussion to #4833.

**Comment by Tezeug** [comment]
I am sure the keys are fine because create_order works with the same keys but it just is not post only.

**Comment by kroitor** [comment]
> I am sure the keys are fine because create_order works with the same keys

Works against the test URL (as you've overrided it) with the same keys?

**Comment by Tezeug** [comment]
Lol, so I have discovered the problem.
It must be 'true' and not True (omg, really DeriBit? :D)... I was trying it with true before but Python has been thinking it is a normal variable... Thank you so much for help. Huge like for you!

**Comment by kroitor** [comment]
@Tezeug oh, great ) Thx for letting us know! ) I'm sure this will be very helpful to others! Really appreciate the feeedback!

**Comment by jDally987** [comment]
Just to clarify, it's because `post_only` (now `postOnly`..whatever) was added a long time ago, and for some reason takes a String. Dunno who botched that one up but it is what it is. If you look at the reduce_only/reduceOnly parameter though, which was added pretty recently (late 2018), they corrected themselves in that case and made it a proper Boolean:

![image](https://user-images.githubusercontent.com/24235975/58116155-9c27ae80-7bc9-11e9-9e05-190d7b185c56.png)

Obviously now it's totally inconsistent and confusing but there ya go, that's why. I just found this out today too myself after spending way too many hours tearing my hair out trying to figure out why none of my orders would go as post-only......nice job deribit.

Moral of the story though is we should all really just be moving to v2 already
